### ISSUE TITLE
Node labels on being moved to other nodes

### ISSUE THREAD

**Issue by fergalk** [kubernetes-sigs/kind#2419]
Kind appears to be applying node labels defined in v1alpha4.Node.Labels onto other nodes in the kind configuration. Other configuration is not moved between nodes, which presents an issue for my use-case (using labels & taints to control pod placement in daemonsets). This behaviour has been observed on mac os & linux - not tested on Windows.

**What happened**:

Labels defined on one worker node were applied to another worker node. The other configuration defined for the node was applied to the correct node.

**What you expected to happen**:

Labels to be applied to correct nodes as per v1alpha4 cluster configuration file.

**How to reproduce it (as minimally and precisely as possible)**:

Create kind config file `config.yaml` with below contents:

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
- role: worker
  labels:
    myLabel: someValue
  extraPortMappings:
  - containerPort: 80
    hostPort: 2000
```

Run `kind create cluster --config config.yaml --name test`

Validate misconfiguration (sample output shown):

```
$ kubectl  get node test-worker -ojsonpath={.metadata.labels}
{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/os":"linux","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"test-worker","kubernetes.io/os":"linux","myLabel":"someValue"}
$ kubectl  get node test-worker2 -ojsonpath={.metadata.labels}
{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/os":"linux","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"test-worker2","kubernetes.io/os":"linux"}
$ docker inspect test-worker -f {{.NetworkSettings.Ports}}
map[]
$ kubecdocker inspect test-worker2 -f {{.NetworkSettings.Ports}}
map[80/tcp:[{0.0.0.0 2000}]]
```

**Anything else we need to know?**: N/A

**Environment:**

- kind version: (use `kind version`): kind v0.11.1 go1.16.4 linux/amd64
- Kubernetes version: (use `kubectl version`): `Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:18:45Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}`
- Docker version: (use `docker info`): 20.10.7
- OS (e.g. from `/etc/os-release`): Ubuntu (also observed on macOS Big Sur)

**Comment by aojea** [comment]
There is no guarantee order for nodes, i.e. the first worker definition is not necessarily `worker` and the second `worker-2`
This is related to the discussion about node names https://github.com/kubernetes-sigs/kind/issues/1401, in Kubernetes cluster the nodes should be referenced or selected by labels , not by the nodename.

**Comment by fergalk** [comment]
Hi @aojea 
I understand, but if you check my ticket you'll see that the portMapping I added ended up on a different node to the labelled node, when they should have been on the same node. I'm not referencing the nodes by node name, the affinity I'm defining in my daemonset uses the label. This is a problem because I need to map a host port to a daemonset pod; this pod is being placed onto a different node due to the incorrect labelling.

**Comment by aojea** [comment]
Ahhh sorry
/Kind bug

**Comment by BenTheElder** [comment]
The comment about guaranteed order is also incorrect. They are named sequentially with the config.

**Comment by aojea** [comment]
> The comment about guaranteed order is also incorrect. They are named sequentially with the config.

oh , I got this totally wrong, sorry ...  let me try again to see if I get it right now

When we process the labels we do it by the ordered list from the config

https://github.com/kubernetes-sigs/kind/blob/1815e007aa280b5fd79b4fbbebeb6adc802e9285/pkg/cluster/internal/create/actions/config/config.go#L106-L114

However, when we apply them to the kubeadm config of each node, 
https://github.com/kubernetes-sigs/kind/blob/1815e007aa280b5fd79b4fbbebeb6adc802e9285/pkg/cluster/internal/create/actions/config/config.go#L125-L152

we use the list returned by docker, that is not ordered AFAIK

https://github.com/kubernetes-sigs/kind/blob/1815e007aa280b5fd79b4fbbebeb6adc802e9285/pkg/cluster/internal/create/actions/config/config.go#L56-L59

**Comment by BenTheElder** [comment]
Yeah, that's a bug in the implementation. We do other per-node config correctly.
